language/move-compiler/src/to_bytecode/optimize/remove_unnecessary_store.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use move_ir_types::ast as IR;
+
+// Removes any unnecessary store/move combo. This could mean moving/copying a value out of a local
+// just to put it back in. Or storing a value in a local just to remove it.
+
+pub fn code(blocks: &mut IR::BytecodeBlocks) -> bool {
+    let mut changed = false;
+    loop {
+        let removed = remove_unnecessary_store(blocks);
+        if !removed {
+            break;
+        }
+        changed = true
+    }
+    changed
+}
+
+fn remove_unnecessary_store(blocks: &mut IR::BytecodeBlocks) -> bool {
```

Comment:
nit: `remove_unnecessary_store` makes me think this is a dead store optimisation (stored but never read), but this is more like a prototypical common sub-expression elimination optimisation, so two questions:

- Should we also optimise dead stores? (Not necessarily in this PR).
- Can we rename this to .... `remove_write_back`, `remove_nop_store`? Naming things is difficult.

There's potential to expand this into a fully-fledged value-numbering optimisation where the instructions don't need to be next to each other, later.

Reply:
> Should we also optimise dead stores? (Not necessarily in this PR).

We mostly do. See eliminate_locals::optimize for the cfgir. I am bad at naming things and tend to forget the names for things that other people use, like dead store. We should maybe rename those to that 

> Can we rename this to .... remove_write_back, remove_nop_store? Naming things is difficult.

Indeed it is! But I like this name and will use it

Reply:
> There's potential to expand this into a fully-fledged value-numbering optimisation where the instructions don't need to be next to each other, later.

This is a bit difficult currently as it requires plumbing through borrow information... I think?